### PR TITLE
Add IP Adapter Style & Composition Modes

### PR DESCRIPTION
**Pull request by blessedcoolant**
## Summary

Until now IP Adapter had complete control on the contents of the output. With this PR, users are now able to select "Style Only" or "Composition Only" to draw just the style or layout of the reference image.

Based off: https://arxiv.org/abs/2404.02733

### New IP Method Option

- `Full` - Both style and layout of the refence image are used.
- `Style Only` - Only the style of the image is used
- `Composition Only` - Only the composition of the image is used.

opera_0BkqZTwObO

### Example Result

demo

### Notes

- Supports both SDXL and SD1.5

### Testing

- Just check and test if it works as expected with all IP Adapter models - both SDXL and SD1.5

## Merge Plan

Good to merge once tested for all edge cases.

### CODE REVIEW

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py [hunk 1]
```diff
@@ -188,11 +208,12 @@ def __call__(
...
+            batch_size, channel, height, width = hidden_states.shape
             hidden_states = hidden_states.transpose(-1, -2).reshape(batch_size, channel, height, width)
 
         if attn.residual_connection:
             hidden_states = hidden_states + residual
 
         hidden_states = hidden_states / attn.rescale_output_factor
 
-        return hidden_states
+        return cast(torch.FloatTensor, hidden_states)
```
Comment (psychedelicious) [inline invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:219]: This block has a comment: `Start unmodified block from AttnProcessor2_0.`

Probably should remove that bc it is now modified

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py [hunk 2]
```diff
@@ -9,6 +10,11 @@
 from invokeai.backend.stable_diffusion.diffusion.regional_prompt_data import RegionalPromptData
 
 
+class IPAdapterAttentionWeights(TypedDict):
+    ip_adapter_weights: List[IPAttentionProcessorWeights]
+    skip: bool
```
Comment (psychedelicious) [inline invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:15]: Nitpick, using a dataclass feels cleaner to me bc you can just use `ip_adapter_attention_weights.ip_adapter_weights` instead of `ip_adapter_attention_weights["ip_adapter_weights"]` but it's not a big deal.

File: invokeai/frontend/web/src/features/metadata/util/parsers.ts [hunk 3]
```diff
@@ -386,6 +386,7 @@ const parseIPAdapter: MetadataParseFunc<IPAdapterConfigMetadata> = async (metada
     clipVisionModel: 'ViT-H',
     controlImage: image?.image_name ?? null,
     weight: weight ?? initialIPAdapter.weight,
+    method: 'full',
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/metadata/util/parsers.ts:389]: Shouldn't we recall this? Same for `clipVisionModel` (missed it in the prev PR)

**Review by psychedelicious — CHANGES_REQUESTED** [review]
Couple small comments, ML logic itself looks fine to me but probably best to get @RyanJDick to review that too.

File: invokeai/frontend/web/src/features/metadata/util/parsers.ts (hunk 3 above)
Comment (blessedcoolant) [reply invokeai/frontend/web/src/features/metadata/util/parsers.ts:389]: Probably should. I added it to the metadata. Wouldn't that retrieve it or is there something I missed?

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py (hunk 2 above)
Comment (blessedcoolant) [reply invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:15]: I'll update that.

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py (hunk 1 above)
Comment (blessedcoolant) [reply invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:219]: Will clean up once Ryan approves the PR.

File: invokeai/frontend/web/src/features/metadata/util/parsers.ts (hunk 3 above)
Comment (psychedelicious) [reply invokeai/frontend/web/src/features/metadata/util/parsers.ts:389]: No it's hardcoded here to 'full', need to use the same pattern as the other IP metadata fields to retrieve it. Also suggest to fall back to `initialIPAdapter.method` instead of e.g. `'full'`

Comment (blessedcoolant) [reply invokeai/frontend/web/src/features/metadata/util/parsers.ts:389]: Oh dang! My bad. I just added the default value there. Didn't realize it wasn't that. Will fix it up.

File: invokeai/backend/stable_diffusion/diffusion/unet_attention_patcher.py [hunk 4]
```diff
@@ -25,10 +33,21 @@ def _prepare_attention_processors(self, unet: UNet2DConditionModel):
...
+                ip_adapter_attention_weights = IPAdapterAttentionWeights(ip_adapter_weights=[], skip=False)
+                for ip_adapter in self._ip_adapters:
+                    ip_adapter_weight = ip_adapter["ip_adapter"].attn_weights.get_attention_processor_weights(idx)
+                    skip = True
+                    for block in ip_adapter["target_blocks"]:
+                        if block in name:
+                            skip = False
+                            break
+                    ip_adapter_attention_weights.ip_adapter_weights = [ip_adapter_weight]
+                    ip_adapter_attention_weights.skip = skip
```
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/diffusion/unet_attention_patcher.py:45]: We should be appending to these values, not overwriting them in each iteration. (See my comment on the `IPAdapterAttentionWeights` type).

**Review by RyanJDick — CHANGES_REQUESTED** [review]
I've found what I think are some significant bugs in the current implementation. I haven't taken a close look at custom_attention.py yet - I expect that fixing these bugs will have lots of trickle-down effects there.

File: invokeai/backend/stable_diffusion/diffusion/unet_attention_patcher.py (hunk 4 above)
Comment (blessedcoolant) [reply invokeai/backend/stable_diffusion/diffusion/unet_attention_patcher.py:45]: Oh man .. That explains why I wasn't getting the correct array length. So the padding wasn't actually needed. Okay. I fixed that up. And now the results are actually even better.

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py [hunk 5]
```diff
@@ -30,23 +37,22 @@ def __init__(
...
+        attention_mask: Optional[torch.Tensor] = None,
+        temb: Optional[torch.Tensor] = None,
+        # For Regional Prompting:
         regional_prompt_data: Optional[RegionalPromptData] = None,
-        percent_through: Optional[torch.FloatTensor] = None,
+        percent_through: Optional[torch.Tensor] = None,
         # For IP-Adapter:
         regional_ip_data: Optional[RegionalIPData] = None,
+        *args,
+        **kwargs,
```
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:55]: Why add *args and **kwargs?

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py [hunk 6]
```diff
@@ -58,7 +64,6 @@ def __call__(
         is_cross_attention = encoder_hidden_states is not None
 
         # Start unmodified block from AttnProcessor2_0.
-        # vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
```
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:61]: Why remove this? This block is still unmodified. Or is this just a style preference thing?

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py [hunk 7]
```diff
@@ -9,6 +10,12 @@
 from invokeai.backend.stable_diffusion.diffusion.regional_prompt_data import RegionalPromptData
 
 
+@dataclass
+class IPAdapterAttentionWeights:
+    ip_adapter_weights: Optional[IPAttentionProcessorWeights]
+    skip: bool
```
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:16]: `ip_adapter_weights` should not be optional

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py [hunk 8]
```diff
@@ -153,29 +159,34 @@ def __call__(
 
                     # Expected ip_hidden_state shape: (batch_size, num_ip_images, ip_seq_len, ip_image_embedding)
 
-                    ip_key = ipa_weights.to_k_ip(ip_hidden_states)
-                    ip_value = ipa_weights.to_v_ip(ip_hidden_states)
+                    if not self._ip_adapter_attention_weights[ipa_index].skip:
+                        if ipa_weights:
```
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:163]: We definitely should not have an `if ipa_weights` condition here (if anything it should have been an `assert ipa_weights is not None`. I'm guessing that this was just added to address a type error, because `IPAdapterAttentionWeights.ip_adapter_weights` was incorrectly marked as optional. We should be able to remove this once that is fixed.

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py [hunk 9]
```diff
@@ -188,11 +199,14 @@ def __call__(
...
 
         if attn.residual_connection:
             hidden_states = hidden_states + residual
 
         hidden_states = hidden_states / attn.rescale_output_factor
+        # End of unmodified block from AttnProcessor2_0
 
-        return hidden_states
+        # casting torch.Tensor to torch.FloatTensor to avoid type issues
+        return cast(torch.FloatTensor, hidden_states)
```
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:212]: It looks like we relaxed all of the input parameter types from`torch.FloatTensor` to `torch.Tensor`, but then added a cast here to make the inferred type more specific. I feel like we should go one way or the other (and it's fine with me if that's `torch.Tensor` everywhere). Is there something more to this that I'm missing?

**Review by RyanJDick — CHANGES_REQUESTED** [review]
Just a few small things to fix, otherwise looks good to me.

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py (hunk 8 above)
Comment (blessedcoolant) [reply invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:163]: The reason it was marked as Optional was because there are instances where we do not pass any. `attn_procs[name] = CustomAttnProcessor2_0()` and then the classes that follow were updated to accommodate that.

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py (hunk 9 above)
Comment (blessedcoolant) [reply invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:212]: It was mainly done to avoid the strict typing checks we have. All our ops on tensors return `torch.Tensor` and not torch.FloatTensor which eventually defaults to `torch.FloatTensor` anyway. So rather than casting everything else, I thought it was just simpler to change everything to `torch.Tensor` and then just cast it to `FloatTensor` in the end to make sure the `__call__` override matches the parent function.

**Comment by blessedcoolant** [comment]
> The InstantStyle paper also suggested using embedding subtraction to isolate the style. Out of curiosity, did you try that at all?

No. Didn't check that. I'll take a look.

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py (hunk 5 above)
Comment (blessedcoolant) [reply invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:55]: So the override function matches the base and we don't get the incompatible override type checking errors in the editor.

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py (hunk 6 above)
Comment (blessedcoolant) [reply invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:61]: Ah. Must have gotten removed accidentally. Was moving around a bunch of code.

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py (hunk 7 above)
Comment (blessedcoolant) [reply invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:16]: Fixed that. This was a trickle down effect from realizing that the list being passed to the main class can be optional. So I thought we can catch this elsewhere but you're right. This shouldn't be optional in the first place. But the other issue still stands. Resolved this.

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py (hunk 8 above)
Comment (blessedcoolant) [reply invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:163]: Actually resolved too.

File: invokeai/backend/stable_diffusion/diffusion/custom_atttention.py (hunk 5 above)
Comment (RyanJDick) [reply invokeai/backend/stable_diffusion/diffusion/custom_atttention.py:55]: Good 'ol python types... I think the behaviour we'd really want here is to raise if anything was passed to `*args` or `**kwargs` since we aren't handling them

**Review by hipsterusername — APPROVED** [review]
